### PR TITLE
[release/2.1] Add portable builds to CI.

### PR DESCRIPTION
**Pull request by crummel**
Skip CI please.

### CODE REVIEW

**Comment by crummel** [comment]
@dotnet-bot test CI please.

**Comment by crummel** [comment]
@dotnet-bot test CI please.

**Comment by crummel** [comment]
@dotnet-bot test CI please.

**Comment by crummel** [comment]
@dotnet-bot test CI please.

**Comment by crummel** [comment]
@dotnet-bot test CI please.

File: netci.groovy [hunk 1]
```diff
@@ -76,44 +76,61 @@ def addBuildStepsAndSetMachineAffinity(def job, String os, String configuration)
...
 {
-  def newJobName = Utilities.getFullJobName(project, "${os}_${configuration}", true);
+  def config = configuration;
+  if (portable) {
+    config = "${configuration}_Portable"
+  }
+  def newJobName = Utilities.getFullJobName(project, "${os}_${config}", true);
   def contextString = "${os} ${configuration}";
+  if (portable) {
+    contextString = "${os} ${configuration} Portable";
```
Comment (dagood) [inline netci.groovy]: `contextString += " Portable";` ?

File: netci.groovy [hunk 2]
```diff
@@ -76,44 +76,61 @@ def addBuildStepsAndSetMachineAffinity(def job, String os, String configuration)
...
   Utilities.setJobTimeout(newJob, 180);
   Utilities.addGithubPRTriggerForBranch(newJob, branch, contextString, triggerPhrase, !runByDefault);
 }
 
-def addPushJob(String project, String branch, String os, String configuration)
+def addPushJob(String project, String branch, String os, String configuration, boolean portable)
 {
     def shortJobName = "${os}_${configuration}";
+    if (portable) {
+      shortJobName = "${os}_${configuration}_Portable";
```
Comment (dagood) [inline netci.groovy]: `+=` here too?

File: netci.groovy [hunk 3]
```diff
@@ -76,44 +76,61 @@ def addBuildStepsAndSetMachineAffinity(def job, String os, String configuration)
...
+// do some but not all the portable builds
+["Ubuntu16.04", "RHEL7.2", "OSX10.12"].each { os ->
+  addPullRequestJob(project, branch, os, "Release", true, true);
+}
+
 // Per push, run all the jobs
 ["Ubuntu16.04", "Fedora24", "Debian8.4", "RHEL7.2", "Windows_NT", "CentOS7.1", "OSX10.12"].each { os ->
   ["Release", "Debug"].each { configuration ->
-    addPushJob(project, branch, os, configuration);
+    [true, false].each { portability ->
```
Comment (dagood) [inline netci.groovy]: Stick with `portable`? Fits better with `true`, `false`, IMO.

File: netci.groovy [hunk 4]
```diff
@@ -76,44 +76,61 @@ def addBuildStepsAndSetMachineAffinity(def job, String os, String configuration)
   setMachineAffinity(job, os);
 }
 
-def addPullRequestJob(String project, String branch, String os, String configuration, boolean runByDefault)
+def addPullRequestJob(String project, String branch, String os, String configuration, boolean portable, boolean runByDefault)
 {
-  def newJobName = Utilities.getFullJobName(project, "${os}_${configuration}", true);
+  def config = configuration;
+  if (portable) {
+    config = "${configuration}_Portable"
```
Comment (dagood) [inline netci.groovy]: `+=` here too?

File: netci.groovy (hunk 1 above)
Comment (crummel) [reply netci.groovy]: Fixed throughout.

File: netci.groovy (hunk 2 above)
Comment (crummel) [reply netci.groovy]: Fixed.

File: netci.groovy (hunk 3 above)
Comment (crummel) [reply netci.groovy]: Agreed, had started with these as strings and missed this one.

File: netci.groovy (hunk 4 above)
Comment (crummel) [reply netci.groovy]: Fixed.

**Comment by crummel** [comment]
One other thing I was considering was adding portable tarball or unshared builds - do we want these?  They're a bit more work to add but I can refactor some.

**Comment by crummel** [comment]
@dotnet-bot test CI please

**Comment by dagood** [comment]
I can't think of a reason setting portable would fail during offline/tarball but not online, but on the other hand it seems like it would be good to know if it doesn't work. Maybe since offline/tarball involves doing an online build first, we could only do an offline build for portable=true on PRs?

**Comment by crummel** [comment]
Hmm, I'm not sure doing a non-portable online build followed by a portable offline build will work - wouldn't we possibly get some prebuilts with the wrong RID?  I'll give this a try.

**Comment by crummel** [comment]
@dotnet-bot test CI please.

**Comment by crummel** [comment]
@dotnet-bot test CI please.

**Comment by crummel** [comment]
@dotnet-bot test CI please.

**Comment by crummel** [comment]
@dotnet-bot test CI please.

**Comment by crummel** [comment]
@dotnet-bot test CI please.

**Comment by crummel** [comment]
@dotnet-bot test CI please.

**Comment by crummel** [comment]
@dotnet-bot test CI please.

**Comment by crummel** [comment]
@dotnet-bot test CI please.

**Comment by dagood** [comment]
When you're done with the CI testing (or merge this) could you let me know? I'd also like to test out https://github.com/dotnet/source-build/pull/776 in practice.

**Comment by crummel** [comment]
I think this part of it is good to merge now, it's just that the new builds will fail until #717 is in, and that's essentially waiting on https://github.com/dotnet/core-eng/issues/4204 now.  I'm fine with merging this now if those builds failing is okay for now.

**Comment by dagood** [comment]
I don't think we should merge this PR in that case, I just want to run the CI generation tests on my PR so you'd have to regenerate again on this PR if you want to test again. I'm going ahead with that, then.

**Comment by crummel** [comment]
Sounds good!

**Comment by crummel** [comment]
@dotnet-bot test CI please.

**Pull request merged**